graphql_compiler/schema_transformation/split_query.py

```diff
@@ -34,6 +39,340 @@ def __init__(self, query_ast):
+    """
+    # Check if there is a split here. If so, split AST, make child query node, return property
+    # field. If not, recurse on all child selections (if any)
+    if isinstance(ast, Field):
+        parent_type_name = type_info.get_parent_type().name
+        edge_field_name = ast.name.value
+        if (parent_type_name, edge_field_name) in edge_to_stitch_fields:
+            # parent_field_name and child_field_name are names of property fields in the stitch
+            parent_field_name, child_field_name = \
+                edge_to_stitch_fields[(parent_type_name, edge_field_name)]
+            # Get parent field with existing directives
+            parent_property_field = _get_property_field(
+                parent_selections, parent_field_name, ast.directives
+            )
+            # Add @output if needed, record out_name
+            parent_output_name = _get_out_name_optionally_add_output(
+                parent_property_field, intermediate_out_name_assigner
+            )
+            # Create child query node around ast
+            child_query_node, child_output_name = _get_child_query_node_and_out_name(
+                ast, type_info, child_field_name, intermediate_out_name_assigner
+            )
+            # Create and add QueryConnections
+            _add_query_connections(
+                query_node, child_query_node, parent_output_name, child_output_name
+            )
+            # Return parent property field used in the stitch, to be added in selections above
+            return parent_property_field
+
+    # No split here
+    if ast.selection_set is None:  # Property field, nothing to recurse on
+        return ast
+    selections = ast.selection_set.selections
+
+    new_selections = []
+    made_changes = False
+
+    type_info.enter(ast.selection_set)
+    for selection in selections:  # Recurse on children
+        type_info.enter(selection)
+        # By the time we reach any cross schema vertex fields, new_selections contains all
+        # property fields, including any new property fields created by previous cross schema
+        # vertex fields, and therefore will not create duplicate new fields
+        new_selection = _split_query_ast_one_level_recursive(
+            query_node, selection, new_selections, type_info, edge_to_stitch_fields,
+            intermediate_out_name_assigner
+        )
+
+        if new_selection is not selection:
+            made_changes = True
+            if is_property_field_ast(new_selection):
+                # If a property field is returned and is different from the input, then this is
+                # a property field used in stitching. If no existing field has this name, insert
+                # the new property field to end of property fields. If some existing field has
+                # this name, replace the existing field with the returned field
+                new_selections = _replace_or_insert_property_field(new_selections, new_selection)
```

Comment:
Might it be easier to do this by just keeping track of the names of property fields we've already seen? This wholesale reassignment of `new_selections` is making me a bit nervous.